docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,883 @@
+Runtime Hook.**
+
+Once a Runtime Hook or a Runtime Hook input/output parameter element has been added to a particular version,
+it can not be removed from that version or have its behavior significantly changed.
+
+**Rule #2 Runtime Hook’s input parameters must be down-convertible, output parameters must be up-convertible.
+Most specifically**
+
+- input parameters must be able to be down-converted from the latest version to previous versions of the same
+  Runtime Hook; this might imply information loss, but the behavior of the previous version of the Runtime Hook
+  must not be affected by this.
+- output parameters must be able to be up-converted from previous versions to current versions of the same
+  Runtime Hook; this means that new information should be nullable or have defaults.
+
+For example assume that we have a Cluster.BeforeUpgrade Runtime Hook with version `v1alpha1` and `v1alpha2`; 
+In order to avoid duplicating code, Cluster API internally will always work at the latest version, `v1alpha2` 
+in the example, but there could be still a deployed Runtime Extension on `v1alpha1`.
+
+This rule makes it possible to call the Runtime Extensions still using the `v1alpha1` by ensuring it is possible
+to down-converting the input parameter for the `v1alpha2` call implemented in CAPI, make the call, and then 
+up-converting the `v1alpha1` output parameter to the v1alpha2 version `CAPI` expects.
+
+**Rule #3: A Runtime Hook version in a given track may not be deprecated until a new version at least as stable
+is released.**
+
+GA Runtime Hook versions can replace GA or beta Runtime Hook versions; beta Runtime Hook versions may not replace
+GA Runtime Hook API versions etc.
+
+**Rule #4: Other than the most recent Runtime Hook versions in each track, older Runtime Hook versions must
+be supported after their announced deprecation for a duration of no less than:
+
+- GA: 12 months or 3 releases (whichever is longer)
+- Beta: 6 months or 2 releases (whichever is longer)
+- Alpha: 0 releases
+  **
+
+### Discovery hook
+
+The Discovery hook must be implemented by all the Runtime Extensions servers, and it is responsible to
+inform the system about the Runtime Extensions it implements.
+
+When invoked the discovery hook is expected to provide the following answer:
+
+```yaml
+status: Success # or Failure
+message: "error message if status == failure"
+items: # Info about implemented runtime extensions
+- name: http-proxy     # Unique name identifying the runtime extension
+  hook:
+    apiVersion: "hook.runtime.cluster.x-k8s.io/v1alpha1"
+    name: "generatePatches"
+  timeoutSeconds: 5    # Default value suggested by the RuntimeExtension developers
+  failurePolicy: Fail  # Default value suggested by the RuntimeExtension developers 
+- ...
+```
+
+Please note the above struct support defining more than one runtime extensions for the same hook, e.g.
+defining more than one "generatePatches" extensions. 
```

Comment:
```suggestion
Please note that the above struct supports defining more than one Runtime Extension for the same hook, e.g.
defining more than one "generatePatches" extensions. 
```

nit, (just because the example currently only defines one hook)